core/controllers/question_test.py

```diff
@@ -111,3 +111,216 @@ def test_get(self):
             '%s/batch' % feconf.QUESTION_DATA_URL, self.payload,
             expect_errors=True)
         self.assertEqual(response.status_int, 404)
+
+
+class QuestionsHandlerTest(test_utils.GenericTestBase):
+    """Test the question handler."""
+
+    def setUp(self):
+        super(QuestionsHandlerTest, self).setUp()
+
+        self.collection_id = 'coll_0'
+        self.exp_id = 'exp_1'
+        self.signup(self.NEW_USER_EMAIL, self.NEW_USER_USERNAME)
+        self.new_user_id = self.get_user_id_from_email(self.NEW_USER_EMAIL)
+
+        # Create a new collection and exploration.
+        self.save_new_valid_collection(
+            self.collection_id, self.new_user_id, exploration_id=self.exp_id)
+
+        # Add a skill.
+        collection_services.update_collection(
+            self.new_user_id, self.collection_id, [{
+                'cmd': collection_domain.CMD_ADD_COLLECTION_SKILL,
+                'name': 'test'
+            }], 'Add a new skill')
+        collection = collection_services.get_collection_by_id(
+            self.collection_id)
+        self.skill_id = collection.get_skill_id_from_skill_name('test')
+        collection_node = collection.get_node(self.exp_id)
+        collection_node.update_acquired_skill_ids([self.skill_id])
+
+        # Update the acquired skill IDs for the exploration.
+        collection_services.update_collection(
+            self.new_user_id, self.collection_id, [{
+                'cmd': collection_domain.CMD_EDIT_COLLECTION_NODE_PROPERTY,
+                'property_name': (
+                    collection_domain.COLLECTION_NODE_PROPERTY_ACQUIRED_SKILL_IDS), # pylint: disable=line-too-long
+                'exploration_id': self.exp_id,
+                'new_value': [self.skill_id]
+            }], 'Update skill')
+
+        self.payload = {}
+
+    def test_post(self):
+        """Test to verify Post method."""
+        question = question_domain.Question(
+            'dummy', 'A Question',
+            exp_domain.State.create_default_state('ABC').to_dict(), 1,
+            self.collection_id, 'en')
+        self.payload['question'] = question.to_dict()
+        self.login(self.NEW_USER_EMAIL)
+        response_json = self.post_json(
+            '%s' % feconf.QUESTION_DATA_URL, self.payload,
+            expect_errors=False)
+        self.assertIn('question_id', response_json.keys())
+
+        del self.payload['question']
+        response = self.testapp.post(
```

Comment:
I know it's weird but this is a fix to make tests pass.
The tests with 200 will only pass if we have self.post_json, self.put_json ... and 404 cases will only pass with testapp. A switch will cause all the tests to fail. I need to dwell deeper to figure out why this is happening.

Reply:
I thought there were expect_errors, expected_status_int args that you can use for the error cases?

Reply:
Yes, I am using them for error cases. I think I am not clear on my issue.
Here is a better explanation:
When I am expecting 404 as status_int for the above line,
I get 404 as status_int when I access the link with testapp but when I access the link using put_json, I get 200 as status_int, which is not equal to 404 and the test fails.

Reply:
That seems odd, and actually unexpected -- I believe we use put_json successfully elsewhere to detect error cases, see e.g. `test_email_functionality_cannot_be_used_by_non_moderators`. I think this needs more investigation.

Did you also use the expect_errors arg when you tested? Were all environmental conditions the same? If you can explain how to reproduce the error (e.g. by swapping out a self.testapp.put for a put_json) then I would be happy to look into it.

Reply:
I tried the following code
<img width="683" alt="screen shot 2018-02-23 at 12 37 58 pm" src="https://user-images.githubusercontent.com/8845039/36616093-685b729e-1897-11e8-9ae5-dd383d2397c4.png">
It gives me this:
<img width="600" alt="screen shot 2018-02-23 at 12 38 34 pm" src="https://user-images.githubusercontent.com/8845039/36616117-72e3489a-1897-11e8-96e5-03a72478d3b6.png">
I tried this:
<img width="604" alt="screen shot 2018-02-23 at 1 08 52 am" src="https://user-images.githubusercontent.com/8845039/36616134-82ba700e-1897-11e8-9e62-ea43b8f4c0e7.png">
I get this:
https://pastebin.com/rgr487gX

Reply:
Not sure why you're passing the dict with a 'payload' key in the second clause. What if you follow the first one and just pass self.payload instead?

Reply:
The dict was used to overcome one of the testapp checks. I have tried removing dict for self.put_json, I get status_int as 200 instead of 404 even though PageNotFoundException is raised.

Reply:
I think you are reading the error message wrongly. If you do not pass in an expected_status_int then the test infrastructure assumes you mean a 200 response. Please make sure to read put_json carefully and look at other examples of its use in the codebase to understand how it works.

I tried the following and it passes:

```
        del self.payload['change_list']
        self.put_json(
            '%s' % feconf.QUESTION_DATA_URL, self.payload, expect_errors=True,
            expected_status_int=404)

        del self.payload['commit_message']
        self.put_json(
            '%s' % feconf.QUESTION_DATA_URL, self.payload, expect_errors=True,
            expected_status_int=404)

        del self.payload['collection_id']
        self.put_json(
            '%s' % feconf.QUESTION_DATA_URL, self.payload, expect_errors=True,
            expected_status_int=404)

        del self.payload['question_id']
        self.put_json(
            '%s' % feconf.QUESTION_DATA_URL, self.payload, expect_errors=True,
            expected_status_int=404)
```

Also I think there is an error with the way this test is written. At each stage you should take the original payload and delete one field from it. Otherwise all the tests might be failing just because you deleted the first change_list field, causing subsequent checks to be irrelevant.

Reply:
Yep, that fixes put_json but when I tried this fix on get_json, I still had issues.
```        
            response = self.get_json(
            '%s/batch' % feconf.QUESTION_DATA_URL,
            expect_errors=True, expected_status_int=404)

        del self.payload['stringified_skill_ids']
        response = self.get_json(
            '%s/batch' % feconf.QUESTION_DATA_URL, self.payload,
            expect_errors=True, expected_status_int=404)
```
I got
``` TypeError: get_json() got an unexpected keyword argument 'expected_status_int' ```
Should I use testapp here?

Reply:
Nope. I suggest looking at the code for get_json() and its other uses in the codebase -- there ought to be other cases which result in 404 errors.

Reply:
(That's how I figured out how to handle put_json, btw.)

Reply:
resolved this for put and post requests but for get requests, it has to do something with the way the response are generated.
I get the following error:
`  File "/Users/giritheja/Desktop/opensource/oppia/core/tests/test_utils.py", line 274, in _parse_json_response
    json_response.content_type, 'application/json')
AssertionError: 'text/html' != 'application/json'`